src/ApiUtls/AuthApiUtls.ts

```diff
@@ -0,0 +1,36 @@
+import { useState } from "react";
+import axios from "axios";
+import { useNavigate } from "react-router-dom";
+import { toast } from "react-toastify";
+import { useDispatch } from "react-redux";
+import { setUserData } from "../Redux/Slices/AuthSlice/AuthSlice";
+
+const baseAuthUrl = `https://upskilling-egypt.com:3005/api/auth`;
+
+const useCustomFetch = () => {
+
+  const dispatch =useDispatch(); 
+  const navigate = useNavigate();
+  const [loading, setLoading] = useState(false);
+
+  const customFetch = (pathUrl: string, data: object, navigateTo: string) => {
+    setLoading(true);
+    axios
+      .post(`${baseAuthUrl}/${pathUrl}`, data)
+      .then((res) => {
+
+        dispatch(setUserData(res.data.data));
+        toast.success(res.data.message);
+        navigate(navigateTo);
+        setLoading(false);
```

Comment:
instead of setting the loading state in both the `then` and  `catch`  blocks, try moving it to the `finally` block. The code in the `finally` block runs after the promise has been resolved whether to success or error.